### PR TITLE
Add onSeen and onInsert to completions

### PR DESCRIPTION
**Pull request by whazor**
*Description of changes:*

Add the `onSeen` and `onInsert` to each completion. This makes it possible to know which completions were effective. 

For inline completions, `onSeen` will only get called when the user selects an option. Otherwise, all completions get called `onSeen` when the autocomplete popup opens.

We call `onInsert` whenever an autocomplete option gets inserted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by akoreman** [comment]
`Autocomplete.inlineEnabled` being true does not mean every result will be inline, some can be hidden using `completer.hideInlinePreview`, so if we want to have different behaviour for inline and non-inline items, we need to have different filtering.

File: src/autocomplete.js [hunk 1]
```diff
@@ -153,13 +153,19 @@ class Autocomplete {
...
-            if (this.popup.isMouseOver && this.setSelectOnHover) { 
+            if (this.popup.isMouseOver && this.setSelectOnHover) {
                 this.tooltipTimer.call(null, null);
                 return;
             }
+        } else if(!hide && this.completions && this.completions.filtered) {
+            for (var i = 0; i < this.completions.filtered.length; i++) {
+                var completion = this.completions.filtered[i];
+                if (completion && completion.onSeen && typeof completion.onSeen === 'function') {
+                    completion.onSeen(this.editor, completion);
```
[the remark this answers is not in the document]
Comment (whazor) [reply src/autocomplete.js:166]: Good call, we should make sure it only gets called once.

**Comment by nightwing** [comment]
How are these supposed to be used, completer items will be created as instances of a class with these methods, or will completer attach separate functions to each item? It may be more efficient to use completer or  completerId and call the method on completer itself, or simply emit event from completer itself and let users listen to it.

Somewhat related https://github.com/ajaxorg/ace/blob/fcd5e707db7d1229c3f29f7576125bb99d8bf472/src/autocomplete.js#L721C9-L721C58 already uses  the second pattern.

File: src/autocomplete.js [hunk 2]
```diff
@@ -145,32 +148,44 @@ class Autocomplete {
         this.stickySelectionTimer.cancel();
         this.stickySelection = false;
     }
-
+    seen(completion) {
+        if (!this.hasSeen.has(completion) && completion && completion.completer && completion.completer.onSeen && typeof completion.completer.onSeen === "function") {
```
Comment (nightwing) [inline src/autocomplete.js:152]: Could you prefix the method name with $ to indicate that it is a private method, and move `has` function call to the end  condition chain. 

Also do we not allow `?.` syntax?

Comment (whazor) [reply src/autocomplete.js:152]: We don't use `?.` anywhere in the code base.

**Comment by whazor** [comment]
@akoreman I changed the behavior for seen with `hideInlinePreview` to be based for non-inline items on whether it is visible in the autocomplete menu.

**Review by akoreman — APPROVED** [review]

**Pull request merged**